frontend/public/components/utils/k8s-watch-hook.ts

```diff
@@ -0,0 +1,248 @@
+import * as React from 'react';
+// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+// @ts-ignore: FIXME missing exports due to out-of-sync @types/react-redux version
+import { useSelector, useDispatch, Dispatch } from 'react-redux';
+import { Map as ImmutableMap } from 'immutable';
+import { createSelectorCreator, defaultMemoize } from 'reselect';
+import { makeQuery, makeReduxID } from './k8s-watcher';
+import * as k8sActions from '../../actions/k8s';
+import { K8sResourceCommon, K8sKind, K8sResourceKindReference, Selector } from '../../module/k8s';
+import { RootState } from '../../redux';
+import { K8sState } from '../../reducers/k8s';
+
+export class NoModelError extends Error {
+  constructor() {
+    super('Model does not exist');
+  }
+}
+
+const getIDandDispatch: GetIDandDispatch = (resource, k8sModel) => {
+  if (!k8sModel || !resource) {
+    return null;
+  }
+  const query = makeQuery(
+    resource.namespace,
+    resource.selector,
+    resource.fieldSelector,
+    resource.name,
+    resource.limit,
+  );
+  const id = makeReduxID(k8sModel, query);
+  const dispatch = resource.isList
+    ? k8sActions.watchK8sList(id, query, k8sModel)
+    : k8sActions.watchK8sObject(id, resource.name, resource.namespace, query, k8sModel);
+  return { id, dispatch };
+};
+
+const getReduxData = (immutableData, resource: WatchK8sResource) => {
+  if (!immutableData) {
+    return null;
+  }
+  if (resource.isList) {
+    return immutableData.toArray().map((a) => a.toJSON());
+  } else if (immutableData.toJSON) {
+    return immutableData.toJSON();
+  }
+  return null;
+};
+
+const useModelsLoaded = (): boolean => {
+  const ref = React.useRef(false);
+  const k8sModels = useSelector<RootState, K8sKind>(({ k8s }) =>
+    k8s.getIn(['RESOURCES', 'models']),
+  );
+  const inFlight = useSelector<RootState, K8sKind>(({ k8s }) =>
+    k8s.getIn(['RESOURCES', 'inFlight']),
+  );
+
+  if (!ref.current && k8sModels.size && !inFlight) {
+    ref.current = true;
+  }
+  return ref.current;
+};
+
+export const useK8sWatchResource = <R extends K8sResourceCommon | K8sResourceCommon[]>(
+  resource: WatchK8sResource,
+): WatchK8sResult<R> => {
+  const modelsLoaded = useModelsLoaded();
+
+  const k8sModel = useSelector<RootState, K8sKind>(({ k8s }) =>
```

Comment:
Maybe we can memoize this as well, via `k8sModelSelectorCreator` like below?

Reply:
I think that would bring any value as this is a simple selector which does not do any computation, it just gets a value from immutable map.

Reply:
Right, since we use ImmutableJS values within the Redux state, relying on referential equality comparison is good enough for simple (so-called "input") selectors which don't perform any derived data computations.